presto-main/src/main/java/com/facebook/presto/sql/gen/BytecodeUtils.java

```diff
@@ -181,46 +179,92 @@ public static BytecodeNode generateInvocation(Scope scope, String name, ScalarFu
         // Index of parameter (without @IsNull) in Presto function
         int realParameterIndex = 0;
 
+        // go through all the choices in the function and then pick the best one
+        List<ScalarImplementationChoice> choices = function.getAllChoices();
+        int bestChoiceIndex = -1;
+        for (int i = 0; i < choices.size(); i++) {
+            boolean isInvalid = false;
+            ScalarImplementationChoice current = choices.get(i);
+            for (int j = 0; j < arguments.size(); j++) {
+                boolean isInputReference = false;
+                boolean isValueType = current.getArgumentProperty(j).getArgumentType() == VALUE_TYPE;
+                if (!isValueType) {
+                    continue;
+                }
+                if (arguments.get(j) instanceof InputReferenceNode) {
+                    isInputReference = true;
+                }
+                NullConvention currentNullConvention = current.getArgumentProperty(j).getNullConvention();
+                if (!isInputReference && currentNullConvention == NullConvention.BLOCK_AND_POSITION) {
+                    isInvalid = true;
+                    break;
+                }
+            }
+            if (!isInvalid) {
+                bestChoiceIndex = i;
+            }
+        }
+
+        checkArgument(bestChoiceIndex >= 0, "None of the choices are valid");
+        ScalarImplementationChoice bestChoice = choices.get(bestChoiceIndex);
+
+        Binding binding = binder.bind(bestChoice.getMethodHandle());
+
+        MethodType methodType = binding.getType();
+        Class<?> returnType = methodType.returnType();
+        Class<?> unboxedReturnType = Primitives.unwrap(returnType);
+
         boolean boundInstance = false;
         while (currentParameterIndex < methodType.parameterArray().length) {
             Class<?> type = methodType.parameterArray()[currentParameterIndex];
             stackTypes.add(type);
-            if (function.getInstanceFactory().isPresent() && !boundInstance) {
-                checkState(type.equals(function.getInstanceFactory().get().type().returnType()), "Mismatched type for instance parameter");
+            if (bestChoice.getInstanceFactory().isPresent() && !boundInstance) {
+                checkState(type.equals(bestChoice.getInstanceFactory().get().type().returnType()), "Mismatched type for instance parameter");
                 block.append(instance.get());
                 boundInstance = true;
             }
             else if (type == ConnectorSession.class) {
                 block.append(scope.getVariable("session"));
             }
             else {
-                block.append(arguments.get(realParameterIndex));
-                ArgumentProperty argumentProperty = function.getArgumentProperty(realParameterIndex);
+                ArgumentProperty argumentProperty = bestChoice.getArgumentProperty(realParameterIndex);
 
                 if (argumentProperty.getArgumentType() == VALUE_TYPE) {
```

Comment:
Do `switch (argumentProperty.getArgumentType())` instead.

Be sure to have a default that throws, just like the switch 2 lines down.